src/lib/paginator/paginator.ts

```diff
@@ -0,0 +1,157 @@
+ * different page size or navigates to another page.
+ */
+export class PageEvent {
+  pageIndex: number;
+  pageSize: number;
+  length: number;
+}
+
+/**
+ * Component to provide navigation between paged information. Displays the size of the current
+ * page, user-selectable options to change that size, what items are being shown, and
+ * navigational button to go to the previous or next page.
+ */
+@Component({
+  moduleId: module.id,
+  selector: 'md-paginator, mat-paginator',
+  templateUrl: 'paginator.html',
+  styleUrls: ['paginator.css'],
+  host: {
+    'class': 'mat-paginator',
+  },
+  providers: [
+    {provide: MATERIAL_COMPATIBILITY_MODE, useValue: false}
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+})
+export class MdPaginator implements OnInit {
+  private _initialized: boolean;
+
+  /** The zero-based page index of the displayed list of items. Defaulted to 0. */
+  @Input() pageIndex: number = 0;
+
+  /** The length of the total number of items that are being paginated. Defaulted to 0. */
+  @Input() length: number = 0;
+
+  /** Number of items to display on a page. By default set to 50. */
+  @Input()
+  get pageSize(): number { return this._pageSize; }
+  set pageSize(pageSize: number) {
+    this._pageSize = pageSize;
+    this._updateDisplayedPageSizeOptions();
+  }
+  private _pageSize: number = 50;
+
+  /** The set of provided page size options to display to the user. */
+  @Input()
+  get pageSizeOptions(): number[] { return this._pageSizeOptions; }
+  set pageSizeOptions(pageSizeOptions: number[]) {
+    this._pageSizeOptions = pageSizeOptions;
+    this._updateDisplayedPageSizeOptions();
+  }
+  private _pageSizeOptions: number[] = [];
+
+  /** Event emitted when the paginator changes the page size or page index. */
+  @Output() page = new EventEmitter<PageEvent>();
+
+  /** Displayed set of page size options. Will be sorted and include current page size. */
+  _displayedPageSizeOptions: number[];
+
+  constructor(public _intl: MdPaginatorIntl) { }
+
+  ngOnInit() {
+    this._initialized = true;
+    this._updateDisplayedPageSizeOptions();
+  }
+
+  /** Increments the page index to the next page index if a next page exists. */
+  nextPage() {
+    if (!this.hasNextPage()) { return; }
+    this.pageIndex++;
+    this._emitPageEvent();
+  }
+
+  /** Decrements the page index to the previous page index if a next page exists. */
+  previousPage() {
+    if (!this.hasPreviousPage()) { return; }
+    this.pageIndex--;
+    this._emitPageEvent();
+  }
+
+  /** Returns true if the user can go to the next page. */
+  hasPreviousPage() {
+    return (this.pageIndex - 1) >= 0 && this.pageSize != 0;
+  }
+
+  /** Returns true if the user can go to the next page. */
+  hasNextPage() {
+    const numberOfPages = Math.ceil(this.length / this.pageSize) - 1;
+    return (this.pageIndex + 1) <= numberOfPages && this.pageSize != 0;
```

Comment:
Hmm, shouldn't `this.pageIndex < numberOfPages` instead of `(this.pageIndex + 1) <= numberOfPages` technically be the same?